GroundDBMemoryPool/src/rdma/util.cc

```diff
@@ -517,12 +415,72 @@ int resources_create(struct resources *res, const char *serverName, const int tc
             ibv_free_device_list(dev_list);
             dev_list = NULL;
         }
-        if (res->sock >= 0)
+    }
+    return rc;
+}
+/******************************************************************************
+ * Function: register_mr
+ *
+ * Input
+ * res pointer to resources structure
+ * buf pointer to the memory segment
+ * size size of the memory segment
+ *
+ * Output
+ * none
+ *
+ * Returns
+ * 0 on success, error code on failure
+ *
+ * Description
+ * Register a new memory region.
+ ******************************************************************************/
+int register_mr(struct resources *res, char* buf, size_t size){
+    int mr_flags = 0;
+    int rc = 0;
+    res->memregs.emplace_back();
+    auto& memreg = res->memregs.back();
+    /* allocate the memory buffer that will hold the data */
+    if (buf == nullptr){
+        memreg.buf = (char *)malloc(size);
+        if (!memreg.buf){
+            fprintf(stderr, "failed to malloc %Zu bytes to memory buffer\n", size);
+            rc = 1;
+            goto register_mr_exit;
+        }
+        memset(memreg.buf, 0, size);
+        memreg.buf_exclusive = true;
```

Comment:
Should make `buf_exclusive` a parameter

Reply:
In my opinion, `buf_exclusive` indicates whether `buf` is allocated during `register_mr()` or points to an existing memory segment. So it seems calculate `buf == nullptr` is sufficient. Correct me if there are other scenarios, or do you have a better name in mind for this variable?

Reply:
Does `buf_exclusive` being true mean many RDMA connections will access this buffer?

I would define 
`char* register_mr(struct resources *res, size_t size, bool buf_exclusive)`, return value is address for new region with NULL indicating failure.

I won't pass existing memory segment as a parameter because it makes `register_mr() `depends on caller's behavior and requires caller to use it in discretion. A likely pitfall is caller allocates `buf` on stack.

Also, how does caller of your register_mr() know what's your new region's address it doesn't return the address.

Reply:
> Does buf_exclusive being true mean many RDMA connections will access this buffer?

It simply means whether the `buf` is created during calling `register_mr()` or not, so that we know whether we should delete it when unregistering this memory region.

> I won't pass existing memory segment as a parameter because it makes register_mr() 's correctness depend on caller's behavior and requires caller to use it in discretion.

I think passing existing memory segment is a necessary option. For example, in our GroundDB design, compute nodes need to call `register_mr()` using its existing local buffer as `buf`.

> A likely pitfall is caller allocates buf on stack.

I don't quite understand this, could you please elaborate?

> Also, how does caller of your register_mr() know what's your new region's address it doesn't return the address.

`res->memregs.back().buf`. But yes, returning the address or index would be better

Reply:
> It simply means whether the `buf` is created during calling `register_mr()` or not, so that we know whether we should delete it when unregistering this memory region.
>> I see. Please rename `buf_exclusive` to "isBufDeletableFlag" please. It is confused with if the buffer can be shared by multiple connections.

> I think passing existing memory segment is a necessary option. For example, in our GroundDB design, compute nodes need to call `register_mr()` using its existing local buffer as `buf`.
> Okay. Would be out of topic, if compute nodes update its local buffer, does it always call rdma_write()  or just do a update locally and perform rdma_write() later on?

Reply:
> Please rename buf_exclusive to "isBufDeletableFlag" please.

Sure, this is way better.

> Would be out of topic, if compute nodes update its local buffer, does it always call rdma_write() or just do a update locally and perform rdma_write() later on?

In our current design, it would be the latter one since we do the dirty page asynchronously. On the other hand, for some synchronous steps, `rdma_write()` should be call right away.

Reply:
Another API `rdma_flush()` is needed.